File: qt/multimedia/gen_qmediametadata.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QMediaMetaData__Key int

const (
	QMediaMetaData__Title              QMediaMetaData__Key = 0
	QMediaMetaData__Author             QMediaMetaData__Key = 1
	QMediaMetaData__Comment            QMediaMetaData__Key = 2
	QMediaMetaData__Description        QMediaMetaData__Key = 3
	QMediaMetaData__Genre              QMediaMetaData__Key = 4
	QMediaMetaData__Date               QMediaMetaData__Key = 5
	QMediaMetaData__Language           QMediaMetaData__Key = 6
	QMediaMetaData__Publisher          QMediaMetaData__Key = 7
	QMediaMetaData__Copyright          QMediaMetaData__Key = 8
	QMediaMetaData__Url                QMediaMetaData__Key = 9
	QMediaMetaData__Duration           QMediaMetaData__Key = 10
	QMediaMetaData__MediaType          QMediaMetaData__Key = 11
	QMediaMetaData__FileFormat         QMediaMetaData__Key = 12
	QMediaMetaData__AudioBitRate       QMediaMetaData__Key = 13
	QMediaMetaData__AudioCodec         QMediaMetaData__Key = 14
	QMediaMetaData__VideoBitRate       QMediaMetaData__Key = 15
	QMediaMetaData__VideoCodec         QMediaMetaData__Key = 16
	QMediaMetaData__VideoFrameRate     QMediaMetaData__Key = 17
	QMediaMetaData__AlbumTitle         QMediaMetaData__Key = 18
	QMediaMetaData__AlbumArtist        QMediaMetaData__Key = 19
	QMediaMetaData__ContributingArtist QMediaMetaData__Key = 20
	QMediaMetaData__TrackNumber        QMediaMetaData__Key = 21
	QMediaMetaData__Composer           QMediaMetaData__Key = 22
	QMediaMetaData__LeadPerformer      QMediaMetaData__Key = 23
	QMediaMetaData__ThumbnailImage     QMediaMetaData__Key = 24
	QMediaMetaData__CoverArtImage      QMediaMetaData__Key = 25
	QMediaMetaData__Orientation        QMediaMetaData__Key = 26
	QMediaMetaData__Resolution         QMediaMetaData__Key = 27
	QMediaMetaData__HasHdrContent      QMediaMetaData__Key = 28
)

type QMediaMetaData struct {
	h          uintptr
	isSubclass bool
}

// NewQMediaMetaData constructs a new QMediaMetaData object.
func NewQMediaMetaData(param1 *QMediaMetaData) *QMediaMetaData {
	g := newQMediaMetaData(QMediaMetaData_new(param1.cPointer()))
	g.isSubclass = true
	return g
}

// NewQMediaMetaData2 constructs a new QMediaMetaData object.
func NewQMediaMetaData2() *QMediaMetaData {
	g := newQMediaMetaData(QMediaMetaData_new2())
	g.isSubclass = true
	return g
}

func (this *QMediaMetaData) Value(k Key) *qt.QVariant {
	_goptr := qt.UnsafeNewQVariant(unsafe.Pointer(QMediaMetaData_Value(this.h, k)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QMediaMetaData) Insert(k Key, value *qt.QVariant) {
	QMediaMetaData_Insert(this.h, k, (*QVariant)(value.UnsafePointer()))
}

func (this *QMediaMetaData) Remove(k Key) {
	QMediaMetaData_Remove(this.h, k)
}

func (this *QMediaMetaData) Keys() []Key {
	var _ma struct_miqt_array = QMediaMetaData_Keys(this.h)
	_ret := make([]Key, int(_ma.len))
	_outCast := (*[0xffff]Key)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		xxxxxxxxx
	}
	return _ret
}

func (this *QMediaMetaData) OperatorSubscript(k Key) *qt.QVariant {
	return qt.UnsafeNewQVariant(unsafe.Pointer(QMediaMetaData_OperatorSubscript(this.h, k)))
}

func (this *QMediaMetaData) Clear() {
	QMediaMetaData_Clear(this.h)
}

func (this *QMediaMetaData) IsEmpty() bool {
	return (bool)(QMediaMetaData_IsEmpty(this.h))
}

func (this *QMediaMetaData) StringValue(k Key) string {
	var _ms struct_miqt_string = QMediaMetaData_StringValue(this.h, k)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QMediaMetaData_MetaDataKeyToString(k Key) string {
	var _ms struct_miqt_string = QMediaMetaData_MetaDataKeyToString(k)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

File: qt/multimedia/gen_qmediametadata.cpp
// +build ignore

#include <QList>
#include <QMediaMetaData>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qmediametadata.h>
#include "gen_qmediametadata.h"
QMediaMetaData* QMediaMetaData_new(QMediaMetaData* param1) {
return new QMediaMetaData(*param1);
}
QMediaMetaData* QMediaMetaData_new2() {
return new QMediaMetaData();
}
QVariant* QMediaMetaData_Value(const QMediaMetaData* self, Key k) {
	return new QVariant(self->value(k));
}
void QMediaMetaData_Insert(QMediaMetaData* self, Key k, QVariant* value) {
	self->insert(k, *value);
}
void QMediaMetaData_Remove(QMediaMetaData* self, Key k) {
	self->remove(k);
}
struct miqt_array /* of Key */  QMediaMetaData_Keys(const QMediaMetaData* self) {
	QList<Key> _ret = self->keys();
	// Convert QList<> from C++ memory to manually-managed C memory
	Key* _arr = static_cast<Key*>(malloc(sizeof(Key) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = _ret[i];
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
QVariant* QMediaMetaData_OperatorSubscript(QMediaMetaData* self, Key k) {
	QVariant& _ret = self->operator[](k);
	// Cast returned reference into pointer
	return &_ret;
}
void QMediaMetaData_Clear(QMediaMetaData* self) {
	self->clear();
}
bool QMediaMetaData_IsEmpty(const QMediaMetaData* self) {
	return self->isEmpty();
}
struct miqt_string QMediaMetaData_StringValue(const QMediaMetaData* self, Key k) {
	QString _ret = self->stringValue(k);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QMediaMetaData_MetaDataKeyToString(Key k) {
	QString _ret = QMediaMetaData::metaDataKeyToString(k);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QMediaMetaData_Delete(QMediaMetaData* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QMediaMetaData*>( self );
	} else {
		delete self;
	}
}
